PixelHunter1995/Inputs/Actions.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System.Collections.Generic;
+using Microsoft.Xna.Framework.Input;
+using System;
+
+namespace PixelHunter1995.Inputs
+{
+    using AllKeys = Either<Keys, MouseKeys>;
+    //using KeyDisjunction = Either<AllKeys, HashSet<AllKeys>>;
+    using KeyDisjunction = Dictionary<Either<Keys, MouseKeys>, SignalState>;
+    using KeyConjunction = HashSet<Dictionary<Either<Keys, MouseKeys>, SignalState>>;
+
+    class Actions : StateMap<Action>
+    {
+
+        private readonly Dictionary<Action, KeyConjunction> binds = new Dictionary<Action, KeyConjunction>();
+        
+        public Actions(Dictionary<Action, KeyConjunction> binds)
+        {
+            this.binds = binds;
+        }
+
+        public void Update(Input input)
+        {
+            var activeActions = new HashSet<Action>();
+
+            foreach (var item in binds)
+            {
+                var action = item.Key;
+                KeyConjunction bind = item.Value;
+
+                bool active = true;
+                foreach (KeyDisjunction disjunction in bind)
+                {
+                    bool any = false;
+                    foreach (var item2 in disjunction)
+                    {
+                        AllKeys key = item2.Key;
+                        SignalState desiredState = item2.Value;
+                        SignalState state = input.GetState(key);
+                        // Current state's edge only matters if an edge is desired.
+                        if (desiredState.IsDown == state.IsDown && (desiredState.IsHeld || state.IsEdge))
```

Comment:
Perhaps it would be easier to understand if you add a compare method for SignalState. And I think you could rename SignalState to something sensible like KeyPress.

Reply:
There is a compare-method, but nothing which is lenient when it comes to the edge like this.
Unsure how worthwhile it would be to add one as this kind of check is only done here, but sure, I'll make a SignalState.NonStrictEquals method and use it instead.
Feel free to give name-suggestions, though I figure this one is rather self-explanatory.

The `SignalState` class is for more things than keypresses (even the "bindable actions" themselves use it as well), and the name serves to disconnect the logic from the concept which is more of a subset of it.
The class SignalState describes a state for something that (if seen continuously) might be described by a digital graph (I often refer to it as 'signal graph', as in, "the graph you get from a digital signal" or "the graph describing the signal of a digital button"), where the word "signal" is something I took from the terminology for https://en.wikipedia.org/wiki/Signal_edge.
If you have ideas for better names feel free to give them though.